Add the usage article for variables/locals/outputs (#543)
hashicorp/web-unified-docs: content/terraform/v1.12.x/docs/language/parameterize/configure.mdx

If a module defines a variable without a `default` argument, Terraform prompts the user to supply a value for that variable before it generates a plan. You can assign variable values in the root module using the following methods:

* [HCP Terraform variables and variable sets](#hcp-terraform)
* The `-var` and `-var-file` options on the [CLI](#command-line-variables)
* [Variable definition files](#variable-definition-files)
* [Environment variables](#environment-variables)

Once you assign a value to a variable, you cannot reassign that variable within the same file. However, if the root module receives multiple values for the same variable name from different sources, Terraform uses the following order of precedence:

1. Any `-var` and `-var-file` options on the [command line](#command-line-variables) in the order provided and variables from HCP Terraform
2. Any `*.auto.tfvars` or `*.auto.tfvars.json` files in lexical order
3. The `terraform.tfvars.json` file
4. The `terraform.tfvars` file
5. Environment variables

> 6. The `default` argument of the `variable` block

Author: Ahh thank you, google docs markdown doesn't change those and I keep forgetting to 🤦‍♀️

- Terraform warns you if you assign an undeclared variable in a [variable definition file](#variable-definition-files), letting you catch accidental misspellings in your configuration or definition files.
- Terraform errors if you attempt to assign a value for an undeclared variable with `-var` on the command line.

## Define locals to reuse expressions

> **Hands-on:** Try the [Simplify Terraform Configuration with Locals](/terraform/tutorials/configuration-language/locals?utm_source=WEBSITE&utm_medium=WEB_IO&utm_offer=ARTICLE_PAGE&utm_content=DOCS) tutorial.

Local values are similar to function-scoped variables in other programming languages. Local values assign names to expressions, letting you use the name multiple times within a module instead of repeating that expression.

Define a `locals` block when you want to reuse an expression throughout your configuration. You can define the `locals` block in any module, and the value you assign can be any valid Terraform expression, and can reference the following:

- Variables
- Resource attributes
- Function outputs
- Other local values

> For example, you could define a `locals` block to create a consistent naming convention, identify your primary subnet, and to set aside environmental configuration settings:
> 
> <CodeBlockConfig hideClipboard>
> 
> ```hcl
> locals {
>   # Naming convention
>   resource_name = "${var.project_name}-${var.environment}"
> 
>   # Process the subnet list
>   primary_public_subnet = var.subnet_ids[0]
>   subnet_count          = length(var.subnet_ids)
> 
>   # Environmental deployment settings
>   is_production      = var.environment == "prod"
>   monitoring_enabled = var.monitoring || local.is_production
> }

Reviewer: ```suggestion
The following example define a `locals` block called `resource_name` to create a consistent naming convention. It also creates a `primary_public_subnet` local and `subnet_count` local for identifying the primary subnet. Finally, it creates an `is_production` local that sets aside environmental configuration settings:

<CodeBlockConfig hideClipboard>

```hcl
locals {
  resource_name = "${var.project_name}-${var.environment}"
  
  primary_public_subnet = var.subnet_ids[0]
  subnet_count          = length(var.subnet_ids)

  is_production      = var.environment == "prod"
  monitoring_enabled = var.monitoring || local.is_production
}
```
I don't feel super strongly about this one, but we would normally avoid describing how something works in the code comments. It it were core documentation and not an example, I'd take a harder stand.

Author: The code comments reflect what someone would actually put in _their_ configuration to help consumers understand their `local` block (I had a comment about this in my draft and forgot to replicate here). So I'm going to ignore this one because I feel it's clear as is

```

</CodeBlockConfig>

> To reference the values from the `locals` block in your configuration, use `local.<NAME>` syntax. The `locals` block defines local values, and you reference those values on an object named `local`, which is singular.

Reviewer: ```suggestion
Use the `local.<NAME>` syntax to reference values from a `locals` block in your configuration. The `locals` block defines local values, but  you must use the singular `local` keyword to reference the individual values.
```
Is that any clearer? Still accurate?

> For example, in your configuration you can reference `local.resource_name` to name resources, and use `local.primary_public_subnet` with `local.monitoring_enabled` to configure a web server:

Reviewer: ```suggestion
In the following example, the tag `Name` argument is set by referencing the value of `local.resource_name`. The configuration sets the subnet ID by referencing `local.primary_public_subnet`. Finally, Terraform retrieves the value of the `local.monitoring_enabled` local to determine if it should enable monitoring for the resource:
```
Ignore this one if you disagree.

<CodeBlockConfig highlight="4,5,8,14,17" hideClipboard>

```hcl
resource "aws_instance" "web" {
  ami           = data.aws_ami.ubuntu.id
  instance_type = var.instance_type
  subnet_id     = local.primary_public_subnet
  monitoring    = local.monitoring_enabled

  tags = {
    Name        = local.resource_name
    Environment = var.environment
  }
}

resource "aws_security_group" "web" {
  name = "${local.resource_name}-sg"

  tags = {
    Name = "${local.resource_name}-security-group"
  }
}
```

</CodeBlockConfig>

You can access local values in the module where you define them, but not in other modules. However, you can pass a local value to a child module as an [argument](/terraform/language/modules/syntax#calling-a-child-module).

> Local values help avoid repetition and give a meaningful name to the value of an expression. However, they can make configuration harder to read because they obscure where values originate. Use local values in situations where you reuse a single value in many places to let you change a value in a single place, or when the value is the result of a complex expression.

Reviewer: ```suggestion
Local values help avoid repetition and give a meaningful name to the value of an expression. However, they can make configuration harder to read because they obscure where values originate. Use local values in situations where you either reuse a single value in many places to let you change a value in a single place or when the value is the result of a complex expression.
```

To learn more about the `locals` block, refer to the [`locals` block reference](/terraform/language/block/locals).

## Define outputs to expose module data

> **Hands-on:** Try the [Output data from Terraform](/terraform/tutorials/configuration-language/outputs) tutorial.

> Outputs let you expose information about your infrastructure on the command line, in HCP Terraform, and in other Terraform configurations. The `output` block serves four main purposes in Terraform:

Reviewer: ```suggestion
Outputs let you expose information about your infrastructure on the command line, in HCP Terraform, and in other Terraform configurations. The `output` block serves the following purposes in Terraform:
```
Just in case it becomes three or five later.

- Child modules can expose resource attributes to parent modules.
- Root modules can display values in CLI output.
- Other Terraform configurations using [remote state](/terraform/language/state/remote) can access root module outputs with the `terraform_remote_state` data source.
- Pass information from a Terraform operation to an automation tool.

> Add `output` blocks to export information about your module. For example, you can add two `output` blocks to expose the ID and the IP address of your configuration's web server:

Reviewer: ```suggestion
Add `output` blocks to export information about your module. For example, you can add two `output` blocks to expose the ID and the IP address of your configuration's web server:
```
I've made similar suggestions elsewhere, but it would be more concrete to describe what's happening "in the following example" and to refer to the names of the arguments and keywords being used to create the examples.

Author: I think folks understand that `instance_id` and `instance_ip` map to " you can add two `output` blocks to expose the ID and the IP address of your configuration's web server" here.

```hcl
module "web_server" {
  source = "./modules/web_server"
  # …
}

resource "aws_route53_record" "web" {
  zone_id = data.aws_route53_zone.main.zone_id
  name    = "web.example.com"
  type    = "A"
  records = [module.web_server.instance_ip]
  #...
}

resource "aws_cloudwatch_alarm" "web_health" {
  alarm_name          = "web-server-health"
  comparison_operator = "GreaterThanThreshold"
  threshold           = "80"

  dimensions = {
    InstanceId = module.web_server.instance_id
  }
  # ...
}
```

</CodeBlockConfig>

If you are outputting sensitive data such as a password or API key, use the `sensitive` argument to prevent Terraform from displaying the value in CLI output:

<CodeBlockConfig highlight="4" hideClipboard>

```hcl
output "database_password" {
  description = "Auto-generated password for the RDS database instance"
  value       = aws_db_instance.main.password
  sensitive   = true
}
```

</CodeBlockConfig>

Terraform does still store the values of sensitive outputs in your state, and if you use the [`terraform output` CLI command](/terraform/cli/commands/output) with the `-json` or `-raw` flags then Terraform displays sensitive outputs in plain text.

> Adding the `ephemeral` argument to an output omits that value from state and plan files all together, but adds restrictions to the values you can assign that output. To learn more about the `ephemeral` argument and handling and outputting sensitive data in your configuration, refer to [Sensitive data in state](/terraform/language/state/sensitive-data).

Reviewer: ```suggestion
Adding the `ephemeral` argument to an output omits that value from state and plan files, but it adds restrictions to the values you can assign that output. To learn more about the `ephemeral` argument and handling and outputting sensitive data in your configuration, refer to [Sensitive data in state](/terraform/language/state/sensitive-data).
```
This might be a better suggestion than the one in line 115.
